### PR TITLE
Add prepared statements support for `Mysql2Adapter`

### PR DESCRIPTION
**Pull request by kamipo**
Reopen #22352 because mysql2 0.4.2 is released.

### CODE REVIEW

**Comment by sgrif** [comment]
@kamipo Looks like this causes an intermittent test failure (possibly order dependent, unsure). I'm going to revert for now, can you look into it? It looks like it might even be a bug in the mysql2 gem. https://travis-ci.org/rails/rails/jobs/93407920

**Comment by kamipo** [comment]
It seems that this is caused by https://github.com/brianmario/mysql2/issues/694. I'll investigating the cause, thanks!

File: activerecord/lib/active_record/connection_adapters/mysql_adapter.rb [hunk 1]
```diff
@@ -407,24 +345,18 @@ def exec_stmt(sql, name, binds, cache_stmt: false)
...
-            metadata.free
-          end
-
-          result_set = ActiveRecord::Result.new(cols, stmt.to_a) if cols
-          affected_rows = stmt.affected_rows
+          result = stmt.result_metadata
+          ret = yield stmt, result
+          result.free if result
 
           stmt.free_result
```
Comment (sodabrew) [inline activerecord/lib/active_record/connection_adapters/mysql_adapter.rb]: Is this freeing the result twice?

Comment (kamipo) [inline activerecord/lib/active_record/connection_adapters/mysql_adapter.rb]: No.

`result.free` calls `mysql_free_result`.
https://github.com/luislavena/mysql-gem/blob/master/ext/mysql_api/mysql.c#L1112-L1121

```c
static VALUE res_free(VALUE obj)
{
    struct mysql_res* resp = DATA_PTR(obj);
    check_free(obj);
    mysql_free_result(resp->res);
    resp->freed = Qtrue;
    store_result_count--;
    return Qnil;
}
```

`stmt.free_result` calls `mysql_stmt_free_result`.
https://github.com/luislavena/mysql-gem/blob/master/ext/mysql_api/mysql.c#L1610-L1618

```c
static VALUE stmt_free_result(VALUE obj)
{
    struct mysql_stmt* s = DATA_PTR(obj);
    check_stmt_closed(obj);
    if (mysql_stmt_free_result(s->stmt))
	mysql_stmt_raise(s->stmt);
    return obj;
}
```

Comment (sodabrew) [inline activerecord/lib/active_record/connection_adapters/mysql_adapter.rb]: Oh, I didn't see your reply before I edited my own comment :) I'd forgotten that the result set and the result metadata are separate items.

**Pull request merged**